### ISSUE TITLE
plotnine?

### ISSUE THREAD

**Issue by trevorcampbell** [UBC-DSCI/introduction-to-datascience-python#166]
Take a look at https://plotnine.readthedocs.io/en/stable/

It seems to be a drop-in replacement for ggplot. Not sure how stable it is yet, but we might want to discuss switching from Altair if we think it's a better long-term plotting pkg.

**Comment by joelostblom** [comment]
Plotnine has been around for a while so it should be quite stable by now, although I haven't used it much myself (I found the online docs not that great when I tried to get into it a while back). In terms of long-term plotting package decisions, my thoughts are that we should move towards packages that support interactivity and are easy to use in e.g. dashboards rather than away from them. So possibly in the future we might switch to something like [Altair in R](https://github.com/vegawidget/altair), but the current limitation there is that companion packages like vl-convert doesn't exist in R yet (similarly for plotnine, there is no ggally in Python).

In general, I find that Vega-Lite/Altair often uses more sane default choices in terms of best practices (e.g. for colors) than ggplot / plotnine. So when I teach Altair and ggplot, I find that there tends more fiddling in ggplot for these reasons, but I am also biased as I knew Altair better and contribute to it (so no big shock what my opinion is here =) ). If there is something that we think is cumbersome in Altair, I'm happy to look into improving it (we recently released a 5.0RC that comes with some syntax improvements, new functionality, and better docs and error messages).

Here is a long answer I wrote on a similar topic recently that also mentions some more details around altair and ggplot (but also other packages). I should also say that DSCI 320 uses Altair, so we might have to sync with that course if we want to switch in the future (or maybe not, I don't know how syncing between courses works at UBC).

<details><summary>Click to view reply</summary>

**High-level versus low-level syntax**
As you said, an advantage of high-level libraries like altair, searborn, and ggplot/plotnine is that the thinking when creating the chart can be kept "data-centric" instead of "graph-centric". I think of visualization as an aid to explore and understand data, and I think this paradigm is really supported by high-level packages. Nowadays they are also so customizable that I can't remember the last time I came across a plotting problem that made me wish I used a low level package instead (I think low-level package like mpl are mostly useful in building high-level or purpose-specific packages).

**Grammar of graphics versus high-level parameter-driven functions**
Altair, ggplot/plotnine, and the new searborn API all have a graphical grammar where you compose charts from individual grammar units just as you compose sentences from words. Seaborn's current syntax and plotly express are more focused on having individual functions for each plot and then that function takes several parameters to control its behavior. For example, the violinplot function has a parameter for whether there should be a strip plot included as well. I find that the grammar of graphics approach makes things more intuitive and it is easier to compose new charts in a predictable manner. This makes working with the library more rewarding to me, and that feeling of "because A and B worked this way, then C should work this way" is so rewarding when it is true. Among the grammar of graphics libraries this happens the most often for me when working with Altair, I think partly because I know it the best, but partly because of the syntax too ([there is some supportive suggestions of that here (under "stats"),](https://app.notascope.io/) showing that Altair specs are generally "close" to each other syntactically, but the work is still preliminary).

**Sensible defaults**
I would also say that Altair has the most sensible default, not only in terms of personal style preferences, but also in terms of actual best practices, e.g. using well-research color schemes and keeping bars the same width and resizing the overall chart size if there are more/less bars overall instead of creating giant bars when there are fewer categories.

**Interactivity**
I saved the best for last =) After working with an interactive viz library, I don't see how I could go back to a static one. From simple things to adding a tooltip to conveniently find out what outlier points are, to adding widgets and selections. Overall this makes me feel that I am always just a few lines from turning my viz into an mini-application that either helps me explore something interactively or that I can share with a friend who doesn't know Python, or publish on a static web page.
Altair is the only visualization package that I know of that has an grammar of interactivity, which allows you to compose interactions between widgets and plots according to similar principles as when creating the plots via the grammar of graphics. This makes the experience of creating plots consistent and gives me increased creativity and flexibility when designing interactions.

Ok that was a lot, I hope it is helpful. To be clear, I don't think Altair is without its drawbacks, mostly in terms of it still missing some common chart types, [which I mention more in detail in this issue](https://github.ubc.ca/ubc-mds/academic/issues/226#issuecomment-21599). You might also find [this answer](https://stackoverflow.com/questions/59845407/plotly-express-vs-altair-vega-lite-for-interactive-plots/66040502#66040502) that I wrote previously helpful.

</details>
